forester/src/nullifier/address/processor.rs

```diff
@@ -94,180 +103,173 @@ impl<T: Indexer<R>, R: RpcConnection> AddressProcessor<T, R> {
-            .await
-            .map_err(|_| ForesterError::Custom("Failed to get address tree proof".to_string()))?;
-        // TODO: use changelog array size from tree config
-        let indexer_changelog = proof.root_seq % ADDRESS_MERKLE_TREE_CHANGELOG;
-        // TODO:
-        // 1. add index changelog current changelog index to the proof or we make them the same size
-        // 2. remove -1 after new zktestnet release
-        let indexer_index_changelog = (proof.root_seq - 1) % ADDRESS_MERKLE_TREE_INDEXED_CHANGELOG;
-
-        debug!("changelog: {:?}", indexer_changelog);
-        debug!("index_changelog: {:?}", indexer_index_changelog);
-
-        let mut retry_count = 0;
-
-        while retry_count < context.config.max_retries {
-            match update_merkle_tree(
-                rpc,
-                &config.payer_keypair,
-                config.address_merkle_tree_queue_pubkey,
-                config.address_merkle_tree_pubkey,
-                address_hashset_index as u16,
-                proof.low_address_index,
-                proof.low_address_value,
-                proof.low_address_next_index,
-                proof.low_address_next_value,
-                proof.low_address_proof,
-                indexer_changelog as u16,
-                indexer_index_changelog as u16,
-            )
-            .await
-            {
-                Ok(true) => {
-                    debug!(
-                        "Successfully updated merkle tree for address: {:?}",
-                        address
-                    );
-                    let mut nullifications = context.successful_nullifications.lock().await;
-                    *nullifications += 1;
-                    info!("Nullifications: {:?}", *nullifications);
-
-                    indexer
-                        .lock()
-                        .await
-                        .address_tree_updated(config.address_merkle_tree_pubkey.to_bytes(), &proof);
-
-                    if *nullifications >= (ADDRESS_MERKLE_TREE_INDEXED_CHANGELOG / 2) as usize {
-                        info!(
-                            "Reached {} successful nullifications. Re-fetching queue.",
-                            *nullifications
-                        );
-                        *nullifications = 0;
-                        drop(nullifications);
-                        return Ok(Some(AddressPipelineStage::FetchAddressQueueData(
-                            context.clone(),
-                        )));
+        // Create a semaphore to limit concurrent tasks
+        // let semaphore = Arc::new(Semaphore::new(100));
+
+        // Spawn tasks for each account_data
+        for account_data in account_data_batch {
+            let tx = tx.clone();
+            let context = context.clone();
+            // let semaphore = semaphore.clone();
+
+            tokio::spawn(async move {
+                // let _permit = semaphore.acquire().await.unwrap();
```

Comment:
leftovers

Reply:
Kind of, yeah, I thought it would probably make sense to put a upper-limit on concurrent threads, but have not decided yet.